File: 25/day25.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
)

func main() {
	card, door, err := parse("input")

	if err != nil {
		os.Exit(1)
	}

	fmt.Println(solve(card, door))
}

func solve(card int, door int) int {
	cardLoop := getLoopSize(card)
	doorLoop := getLoopSize(door)
	cardEncrypt := getEncryptionKey(card, doorLoop)
	doorEncrypt := getEncryptionKey(door, cardLoop)
	if cardEncrypt == doorEncrypt {
		return cardEncrypt
	}
	return -1
}

func getLoopSize(n int) int {
	v := 1
	s := 7
	loop := 0
	for loop = 0; v != n; loop++ {
		v = (v * s) % 20201227
	}
	return loop
}

func getEncryptionKey(s int, l int) int {
	v := 1
	for i := 0; i < l; i++ {
		v = (v * s) % 20201227
	}
	return v
}

func parse(path string) (int, int, error) {
	card := -1
	door := -1
	file, err := os.Open(path)

	if err != nil {
		return card, door, err
	}

	defer file.Close()

	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		n, err := strconv.Atoi(scanner.Text())

		if err != nil {
			return -1, -1, err
		}

		if card == -1 {
			card = n
		} else {
			door = n
		}
	}

	return card, door, err
}

File: 25/day25.py
#!/usr/bin/env python3


def main():
    with open("input", "r") as f:
        card, door = map(int, f.read().strip().splitlines())
    print(solve(card, door))


def solve(card, door):
    card_loop = get_loop_size(card)
    door_loop = get_loop_size(door)
    card_encrypt = get_encryption_key(card, door_loop)
    door_encrypt = get_encryption_key(door, card_loop)
    return card_encrypt if card_encrypt == door_encrypt else -1


def get_loop_size(n):
    v = 1
    s = 7
    loop = 0
    while v != n:
        v = (v * s) % 20201227
        loop += 1
    return loop


def get_encryption_key(s, l):
    v = 1
    for i in range(l):
        v = (v * s) % 20201227
    return v


if __name__ == "__main__":
    main()
